File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/bluetooth/RadioArbiter.kt
package dev.bota.sdk.internal.bluetooth

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

internal enum class RadioPriority { BackgroundReconnect, ManualSelection }

internal data class RadioOwner(val peripheralId: String, val priority: RadioPriority)

internal class RadioArbiter {
    private val mutex = Mutex()
    private var currentOwner: RadioOwner? = null

    suspend fun owner(): RadioOwner? = mutex.withLock { currentOwner }

    suspend fun acquire(peripheralId: String, priority: RadioPriority): String? = mutex.withLock {
        val owner = currentOwner
        when {
            owner == null -> {
                currentOwner = RadioOwner(peripheralId, priority)
                null
            }
            owner.peripheralId == peripheralId -> {
                currentOwner = RadioOwner(
                    peripheralId,
                    if (priority.ordinal > owner.priority.ordinal) priority else owner.priority,
                )
                null
            }
            priority.ordinal > owner.priority.ordinal -> {
                currentOwner = RadioOwner(peripheralId, priority)
                owner.peripheralId
            }
            else -> peripheralId
        }
    }

    suspend fun release(peripheralId: String) = mutex.withLock {
        if (currentOwner?.peripheralId == peripheralId) currentOwner = null
    }
}

File: platforms/apple/Sources/BotaAppSDK/Bluetooth/RadioArbiter.swift
enum RadioPriority: Int, Sendable {
    case backgroundReconnect = 0
    case manualSelection = 1
}

struct RadioOwner: Equatable, Sendable {
    let peripheralID: String
    let priority: RadioPriority
}

actor RadioArbiter {
    private(set) var owner: RadioOwner?

    func acquire(peripheralID: String, priority: RadioPriority) -> String? {
        guard let owner else {
            self.owner = RadioOwner(peripheralID: peripheralID, priority: priority)
            return nil
        }
        if owner.peripheralID == peripheralID {
            self.owner = RadioOwner(peripheralID: peripheralID, priority: max(owner.priority, priority))
            return nil
        }
        guard priority.rawValue > owner.priority.rawValue else { return peripheralID }
        self.owner = RadioOwner(peripheralID: peripheralID, priority: priority)
        return owner.peripheralID
    }

    func release(peripheralID: String) {
        if owner?.peripheralID == peripheralID { owner = nil }
    }
}

private func max(_ lhs: RadioPriority, _ rhs: RadioPriority) -> RadioPriority {
    lhs.rawValue >= rhs.rawValue ? lhs : rhs
}
